File: python/examples/basic_usage.py
"""Basic usage: create a sandbox, run commands, transfer files, clean up.

Prerequisites:
  - agent-sandbox controller and CRDs installed in your cluster
  - sandbox-router deployed
  - SandboxTemplate "python-sandbox-template" applied:
      kubectl apply -f examples/k8s/sandbox-template.yaml

Run:
  python examples/basic_usage.py
"""

from __future__ import annotations

from langchain_kubernetes import KubernetesProvider, KubernetesProviderConfig

config = KubernetesProviderConfig(
    template_name="python-sandbox-template",
    namespace="default",
    connection_mode="tunnel",  # uses kubectl port-forward automatically
    startup_timeout_seconds=120,
)

provider = KubernetesProvider(config)

# Create a new sandbox (provisions a Sandbox CR, waits for readiness)
sandbox = provider.get_or_create()
print(f"Sandbox created: {sandbox.id}")

try:
    # Run a shell command
    result = sandbox.execute("echo 'Hello from Kubernetes!'")
    print(f"Output: {result.output!r}  exit_code={result.exit_code}")

    # Run a Python snippet
    result = sandbox.execute("python3 -c 'import sys; print(sys.version)'")
    print(f"Python version: {result.output.strip()}")

    # Upload a file
    upload_resp = sandbox.upload_files([("/tmp/demo.txt", b"Hello, DeepAgents!\n")])
    print(f"Upload: {upload_resp[0].path!r}  error={upload_resp[0].error}")

    # Download the file back
    dl_resp = sandbox.download_files(["/tmp/demo.txt"])
    print(f"Downloaded {len(dl_resp[0].content or b'')} bytes from /tmp/demo.txt")

    # Use BaseSandbox filesystem helpers (implemented via execute())
    sandbox.execute("mkdir -p /tmp/project && echo 'x=1' > /tmp/project/script.py")
    entries = sandbox.ls_info("/tmp/project")
    print(f"ls /tmp/project: {[e['path'] for e in entries]}")

    # List active sandboxes managed by this provider instance
    active = provider.list()
    print(f"Active sandboxes: {[s.id for s in active]}")

finally:
    provider.delete(sandbox_id=sandbox.id)
    print("Sandbox deleted.")

File: typescript/examples/basic-usage.ts
/**
 * Basic usage example for langchain-kubernetes.
 *
 * Prerequisites:
 *   - A running Kubernetes cluster accessible via the default kubeconfig
 *   - The "deepagents-sandboxes" namespace (or let the provider create it)
 *
 * Run:
 *   npx tsx examples/basic-usage.ts
 */
import { KubernetesProvider } from "../src/index.js";

async function main() {
  const provider = new KubernetesProvider({
    image: "python:3.12-slim",
    blockNetwork: false, // enable network for this demo
    startupTimeoutSeconds: 120,
  });

  console.log("Creating sandbox…");
  const sandbox = await provider.getOrCreate();
  console.log(`Sandbox ready. id=${sandbox.id}`);

  // ── Execute a simple command ──────────────────────────────────────────────
  let result = await sandbox.execute("python3 -c \"print('Hello from Kubernetes!')\"");
  console.log("stdout:", result.output.trim());
  console.log("exitCode:", result.exitCode);

  // ── Write and read a file via the inherited BaseSandbox helpers ───────────
  result = await sandbox.execute("echo 'Hello, file!' > /workspace/greeting.txt");
  result = await sandbox.execute("cat /workspace/greeting.txt");
  console.log("file content:", result.output.trim());

  // ── Upload a file via tar transport ───────────────────────────────────────
  const uploadResults = await sandbox.uploadFiles([
    ["/workspace/script.py", Buffer.from("print('Uploaded script ran!')\n")],
  ]);
  console.log("upload success:", uploadResults[0]?.success);

  result = await sandbox.execute("python3 /workspace/script.py");
  console.log("script output:", result.output.trim());

  // ── Download a file via tar transport ─────────────────────────────────────
  const downloadResults = await sandbox.downloadFiles(["/workspace/greeting.txt"]);
  console.log(
    "downloaded:",
    Buffer.from(downloadResults[0]!.content).toString().trim()
  );

  // ── Cleanup ───────────────────────────────────────────────────────────────
  console.log("Deleting sandbox…");
  await provider.delete(sandbox.id);
  console.log("Done.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
